book/src/java_objects_in_rust.md

```diff
@@ -0,0 +1,75 @@
+# Java Objects in Rust
+
+At this point, suppose we’ve concluded that we need to run `counter_increment`
+on a large number of objects at once. We might add an API to the Java side like
+so:
+
+
+```java
+class NativeAPI {
+    ...
+    static native void increment_all(NativeCounter[] counters)
+}
+```
+
+On the Rust side, our corresponding signature will be:
+
+```rust
+#[no_mangle]
+pub extern "system" fn Java_com_github_jni_1rs_jnibook_NativeAPI_increment_all(
+    env: JNIEnv,
+    _class: JClass,
+    counters: jobjectArray,
+) {
+    let length = env.get_array_length(pointers)
+        .expect("TODO: proper error handling");
+    for index in 0..length {
+        let item = env.get_object_array_element(counters)
+            .expect("TODO: proper error handling");
+        let pointer = env.get_field(item, "ptr", "Z")
+            .expect("TODO: proper error handling");
+        // TODO: Obtain the Rust object and invoke it.
+    }
+}
+```
+
+## Exercises
+
+1. Fill in the code to invoke `increment()` on the underlying Rust object.
+2. If this code is run with a large enough object array, it will crash. That’s
+   because the call to `get_object_array_element` creates a local reference, and
+   only a small number of local references are permitted at a time (the exact
+   number may vary based on your OS). Fix the code using one of the following
+   approaches:
+    1. Releasing the local reference when you’re done with it, using
+       [`delete_local_ref`](https://docs.rs/jni/0.15.0/jni/struct.JNIEnv.html#method.delete_local_ref),
+       as you might in C++.
+    2. Give each iteration its own local reference frame using
+       [`with_local_frame`](https://docs.rs/jni/0.15.0/jni/struct.JNIEnv.html#method.with_local_frame),
+       so that all references within that frame are released by the next
+       iteration.
+    3. Use the
+       [`AutoLocal`](https://docs.rs/jni/0.15.0/jni/struct.JNIEnv.html#method.auto_local)
+       helper implemented within the JNI crate, so you don’t have to manually
+       track the usage of that reference at all.
+
+## Notes
+
+The
+[`AutoLocal`](https://docs.rs/jni/0.15.0/jni/struct.JNIEnv.html#method.with_local_frame)
+wrapper uses Rust’s version of RAII (Resource Acquisition Is Initialization) to
+call `delete_local_ref` when the object goes out of scope. By relying on the
+compiler’s ownership and borrowing checks, AutoLocal safely tracks the local
+usage of your objects and prevents you from releasing a reference too early. How
+does it do that? `AutoLocal` takes ownership of the reference when you
+initialize it, so you can’t have any other references to one (without using
+unsafe code). The method `AutoLocal.as_obj()` will hand out `JObject` references
+*only as long as* those references are shorter-lived than the `AutoLocal`
+itself.
+
+This is a contrived example (we would probably convert to `long` on the Java
```

Comment:
My concern is the same: we show an example that doesn't seem to make sense
(Java loop over NativeCounters will be faster, because in that case no "Rust -> Java"
calls will happen; passing an array of longs may or may not be faster than a Java loop, but certainly will be faster than the example).

Maybe passing an array of Points to print/render them? Again, Points as Java objects might not be the most efficient solution in such a case :)

Reply:
Part of the motivation is to familiarize the reader with arrays of objects, and I agree with you that the solution doesn't make much sense for the problem considering all the extra Rust->Java calls. On the other hand, I like that it introduces field and array access.

We could probably replace NativeCounter with a different class (I'm open to it) or perform a different operation. I'm partly tempted by your suggestion to pass an array of Points or longs, since we could maybe move it to construction time (potentially not the critical path) and talk a little bit better about the cost of the JNI upcalls.

Reply:
> Again, Points as Java objects might not be the most efficient solution in such a case :)

Possibly, some objects with behaviour (not plain data holders, as with Points readers could see more efficient approaches), e.g., Java passing to native an array of `Spaceship`s, and native `Spaceship#launch`ing them.

Reply:
For an outline of this chapter, how do you feel about:

- Problem description: Describe as an exercise for updating many X (possibly Counters)
- Solution 1: Java loop. Describe as simplest solution.
- Solution 2: Rust loop with ptrs. Describe as more complex, but potentially warranted if performance tests demonstrated an improvement.
- Solution 3: Rust loop with Java objects. Describe as most complex, with the highest number of upcalls. Mention that in certain cases, JNI performance considerations may be negligible (As in , `Spaceship#launch`).
- Analysis: Summarize tradeoffs for each solution. (Number of upcalls/downcalls in each, complexity, additional code, flexibility for future changes). Conclude with an author's preference for 1 based off simplicity. Note that which of Solution 1 and 2 is faster depends on the JNI implementation, and list some considerations (for example, array copying).

As a result, the reader should get exposed to a few different patterns for managing arrays across Rust and Java, and they would be more inclined to reach for Solution 1, or try Solution 2 after testing performance themselves. However, they'd also be able to get data out of a Java object and LongArrays if they needed to.

Reply:
Depends on what this chapter attempts to cover. Is the focus on using Java objects in Rust, or accessing Java arrays?

For instance, if the focus is on passing ~~Java objects~~ data, I think we can make an example with a single argument, and consider two "types" of objects the native might need:
- Plain data carriers. In this case, accessing Java objects may or may not be the best idea (some of the considerations are the speed of development and the runtime efficiency). There are at least three viable approaches:
  - Pass Java objects as-is. Native -> Java method invocations are rather expensive; don't have data on field accesses. However, sometimes it might be simpler than other approaches in some cases.
  - Unroll a Java object into a collection of primitive fields (e.g., instead of passing a single `3dPoint`, pass `x, y, z`).
  - Pass the Java object in a serialized form as a byte array, e.g., a protocol buffer. Depending on the number and types of fields and serde performance, might or might not be better.
- Objects with behaviour that Rust _needs_ to call (e.g., Java object implements a Rust trait). In this case, no much choice here: you have to pass a `Spaceship` if the `MissionControlCenter` is in Rust.

If we limit tje discussion to objects (and do not discuss passing data), there will be only one solution; if extend to data: at least three.

I'd consider keeping arrays as a separate sub-section, if we discuss passing data.

Reply:
Based off this feedback, I've introduced arrays in a separate exercise (together with string passing). (Actually, the whole thing is kind of rewritten, to make it less wordy and more about completing exercises.) I think the discussion in the chart exercise should also be updated to cover object unrolling, and plan on introducing native->Java performance separately (maybe as a profiler lab, so that people can try it themselves).